ZiLing/Midterm-Review.py

```diff
@@ -0,0 +1,481 @@
+
++ - addition
+- - subtraction 
+* - multiplication
+/ - division 
+% - Modules
+** - Exponentiation
+// = Floor Division
+
+Remember that there are two types of numbers 
+int: Intergers, whole numbers 
+float: Decimal numbers 
+
+if you want more math, you have to use the math library 
+
+import math 
+
+then you can do all your wacky math.cos(),math.sine(),math.sqrt() stuff 
+
+Down there are the example of some 
+"""
+# Module is the remain after division
+# for example, 2/2 is 1, there is no remainder
+# 3/2 can also be see as 2+1/2, the remainder will be 1 
+print(2%2)
+print(3%2)
+
+# be aware of the difference in decimal places
+print(2//3)
+print(2/3)
+
+# you can also use + and * on string 
+word = "hello" + " world"
+hohoho = 'ho'*3
+
+"""
+Assignment Operators (in-place operators)
+You dont have to use them, but just know what they are 
+
+= -> x = 5 
++= -> x += 3 is the same sa x = x + 3 
+-= -> x -= 3 is the same as x = x - 3 
+*= -> x *= 3 is the same sa x = x * 3 
+/= -> x /= 3 is the same sa x = x / 3 
+%= -> x %= 3 is the same sa x = x % 3 
+**= -> x **= 3 is the same sa x = x ** 3 
+
+there are more than these, but just know the onces above should be enough 
+for you to go through this course
+
+"""
+
+# You cannot use two assignment operators on the same variable at the same line 
+# But you can mix it with other algebretical operators 
+x += 435+493/2302567
+
+"""
+Everything below will return a boolean value(True or False)
+
+Comparison Oprators 
+
+Operator   Name           Example
+==         Equal          x == y, 2 == 1 
+!=       Not Equal       x != 7, y != y
+>      Greater than       3 > 2, 4 > 5
+<       Less than         1 < 2, 5 < 5 
+>=   Greater or Equal to   1 >= 1, 1 >= 4
+<=   less or equal to      x <= 3 , 4 <= 4 
+
+You usually use them with if statement or while statement
+"""
+
+# You can use more than one comparison operators in the same line 
+for i in range (356,9,-3): 
+    if 325 < i > 35:
+        print (i)
+
+"""
+Logical Oprator 
+
+operator    Returns True if          Example
+and     both statements are true     x < 5 and x > 1 
+or      either statement is true     x < 5 or x > 1 
+not     the statement is false       not(x > 5)
+
+Identity Operator 
+
+operator    Returns True if          Example
+is      both variables are the same   5 is 5, x is 5
+is not  variables are not the same    x is not 4 
+
+Membership Operator 
+
+operator    Returns True if                 Example
+in       object is element in smt        i in mydict
+not in      object not in smth           i not in mydict     
+
+All the operators above return a boolean type value (True or False)
+"""
+
+# You can mix mash all the operators together as long as it makes sense
+for i in range (1,101):
+    if not(x%2 == 0 and x > 5 ): 
+        print (i)
+
+"""
+Conditional statements 
+
+Statement      name   
+if              if      
+elif         else if
+else           else
+while          while 
+"""
+
```

Comment:
Note that while-loops should be avoided where possible. They are coded natively in pure python, and are very slow. In contrast, a for-loop are coded natively in C. Here's a test you can run to see the difference between two functions that iterate over a list.
```python
from time import perf_counter

testList = [x for x in range(100_000_000)]

def while_loop_iterator(list: list):
    """
    Test function that returns the last element of a python list and prints the time it takes to iterate through the
    list using a while-loop

    :param list: a non-empty list of integers to iterate through
    :return value: The final element in the list
    """

    start = perf_counter()
    index = 0
    maxIndex = len(list)
    while index < maxIndex:
        value = list[index]
        index += 1
    end = perf_counter()
    print(end-start)
    return value

def for_loop_iterator(list: list):
    """
        Test function that returns the last element of a python list and prints the time it takes to iterate through
        the list using a for-loop

        :param list: a non-empty list of integers to iterate through
        :return value: The final element in the list
        """

    start = perf_counter()
    for i in range(0,len(list)):
        value = list[i]
    end = perf_counter()
    print(end-start)
    return value

if __name__ == "__main__":
    while_loop_iterator(testList)
    for_loop_iterator(testList)
```

Something else to keep in mind is that `for i in range(0,len(list))` generates the range once and iterates through that range. Thus, the following two ways of implementing the loop are equivalent:
```python
iterableRange = range(0, len(list))
for i in iterableRange:
```
```python
for i in range(0, len(list))
```

While loops in comparison will generate the variable you are using for the comparison every time if it is not pre-defined. Thus, the following two methods are not equivalent, and the second method takes far longer:
```python
maxIndex = len(list)
index = 0
while index < maxIndex:
```
```python
index = 0
while index < len(list):
```

Lastly, comprehension is even faster with for-loops. This is because if you define an index specifically, a comparison must be done with your explicit index and the index of an object stored in an array, or it must generate the index if you are iterating through something like a C-style list. This comparison is being done with an object coded in pure Python. Comprehension uses built-in methods for defining the position of elements in an iterable and is coded in C. The following function is even faster:
```python
 def for_loop_iterator_comprehsion(list: list):
    """
        Test function that returns the last element of a python list and prints the time it takes to iterate through
        the list using a for-loop and list comprehension

        :param list: a non-empty list of integers to iterate through
        :return value: The final element in the list
        """

    start = perf_counter()
    for element in list:
        value = element
    end = perf_counter()
    print(end-start)
    return value
```